packages/shared_preferences/shared_preferences/lib/shared_preference_async.dart

```diff
@@ -0,0 +1,313 @@
+  SharedPreferencesAsync({required SharedPreferencesOptions options})
+      : _options = options {
+    if (SharedPreferencesAsyncPlatform.instance == null) {
+      throw PlatformException(
+          code: 'The SharedPreferencesAsyncPlatform instance must be set.');
+    } else {
+      _platform = SharedPreferencesAsyncPlatform.instance!;
+    }
+  }
+
+  /// Options that determine the behavior of contained methods,  usually
+  /// platform specific extensions of the [SharedPreferencesOptions] class.
+  final SharedPreferencesOptions _options;
+
+  late final SharedPreferencesAsyncPlatform _platform;
+
+  /// Returns all keys on the the platform that match provided [parameters].
+  ///
+  /// If no restrictions are provided, fetches all keys stored on the platform.
+  ///
+  /// Ignores any keys whose values are types which are incompatible with shared_preferences.
+  Future<Set<String?>> getKeys(GetPreferencesParameters parameters) async =>
+      _platform.getKeys(parameters, _options);
+
+  /// Returns all keys and values on the the platform that match provided [parameters].
+  ///
+  /// If no restrictions are provided, fetches all entries stored on the platform.
+  ///
+  /// Ignores any entries of types which are incompatible with shared_preferences.
+  Future<Map<String, Object?>> getAll(
+      GetPreferencesParameters parameters) async {
+    return _platform.getPreferences(parameters, _options);
+  }
+
+  /// Reads a value from the platform, throwing a [TypeError] if the value is not a
+  /// bool.
+  Future<bool?> getBool(String key) async => _platform.getBool(key, _options);
+
+  /// Reads a value from the platform, throwing a [TypeError] if the value is not
+  /// an int.
+  Future<int?> getInt(String key) async => _platform.getInt(key, _options);
+
+  /// Reads a value from the platform, throwing a [TypeError] if the value is not a
+  /// double.
+  Future<double?> getDouble(String key) async =>
+      _platform.getDouble(key, _options);
+
+  /// Reads a value from the platform, throwing a [TypeError] if the value is not a
+  /// String.
+  Future<String?> getString(String key) async =>
+      _platform.getString(key, _options);
+
+  /// Reads a list of string values from the platform, throwing an
+  /// exception if it's not a string list.
+  Future<List<String>?> getStringList(String key) async {
+    List<dynamic>? list = await _platform.getStringList(key, _options);
+    if (list is List<String?>) {
+      list = list.cast<String>().toList();
+    }
+    // Make a copy of the list so that later mutations won't propagate
+    return list?.toList() as List<String>?;
+  }
+
+  /// Returns true if the the platform contains the given [key].
+  Future<bool> containsKey(String key) async =>
+      (await getKeys(GetPreferencesParameters(
+              filter: PreferencesFilters(allowList: <String>{key}))))
+          .isNotEmpty;
+
+  /// Saves a boolean [value] to the platform.
+  Future<bool> setBool(String key, bool value) =>
```

Comment:
What do all of the `bool` return values mean? It should be documented.

Reply:
(Alternatively, if we don't think it's something that should happen in practice, they could be `void` and throw exceptions for any crazy errors.)